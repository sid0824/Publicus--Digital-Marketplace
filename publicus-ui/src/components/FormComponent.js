import React from "react";
import { InputTypes } from "../pages/CreateForm";
import {
  DatePicker,
  Form,
  Input,
  InputNumber,
  Radio,
  Select,
  Switch,
  TimePicker
} from "antd";
const { Option } = Select;

// The types of string inputs currently available
export const ruleTypes = ["email", "number", "url"];

// Converts the list of supplied rules into a format that the form item can understand and apply
const generateFormItemRules = rules => {
  const formattedRules = [];
  rules.forEach(rule => {
    if (ruleTypes.includes(rule)) {
      formattedRules.push({ type: rule, message: `Value must be a ${rule}` });
    }
    if (rule === "required") {
      formattedRules.push({
        required: true,
        message: `This field is compulsory`
      });
    }
  });
  return formattedRules;
};

/**
 * A generic form component used in both the complete form and create form pages. It renders a form item
 * based on the configuration given with the 'question' prop for a form belonging to the 'form' prop.
 */
class FormComponent extends React.Component {
  // Return the relevant form component based on the supplied props
  getFormComponent = () => {
    const { id, question, form } = this.props;
    const { getFieldDecorator } = form;
    switch (question.type) {
      case InputTypes.TextInput:
        return (
          <Form.Item>
            {getFieldDecorator(id, {
              rules: generateFormItemRules(question.rules)
            })(<Input style={{ width: 500 }} />)}
          </Form.Item>
        );
      case InputTypes.Select:
        return (
          <Form.Item>
            {getFieldDecorator(id, {
              rules: generateFormItemRules(question.rules)
            })(
              <Select style={{ width: 500 }}>
                {question.options.map(option => (
                  <Option key={option}>{option}</Option>
                ))}
              </Select>
            )}
          </Form.Item>
        );
      case InputTypes.NumberInput:
        return (
          <Form.Item>
            {getFieldDecorator(id, {
              rules: generateFormItemRules(question.rules),
              initialValue: question.startingNumber
            })(
              <InputNumber
                style={{ width: 500 }}
                step={question.incrementor}
                min={question.min}
                max={question.max}
              />
            )}
          </Form.Item>
        );
      case InputTypes.Switch:
        return (
          <Form.Item>
            {getFieldDecorator(id, {
              rules: generateFormItemRules(question.rules)
            })(<Switch />)}
          </Form.Item>
        );
      case InputTypes.TimePicker:
        return (
          <Form.Item>
            {getFieldDecorator(id, {
              rules: generateFormItemRules(question.rules)
            })(<TimePicker />)}
          </Form.Item>
        );
      case InputTypes.DatePicker:
        return (
          <Form.Item>
            {getFieldDecorator(id, {
              rules: generateFormItemRules(question.rules)
            })(
              <DatePicker
                {...(question.showTime
                  ? { showTime: true, format: "YYYY-MM-DD HH:mm:ss" }
                  : {})}
              />
            )}
          </Form.Item>
        );
      case InputTypes.Radio:
        return (
          <Form.Item>
            {getFieldDecorator(id, {
              rules: generateFormItemRules(question.rules)
            })(
              <Radio.Group>
                {question.radioOptions.map(option => (
                  <Radio value={option}>{option}</Radio>
                ))}
              </Radio.Group>
            )}
          </Form.Item>
        );
      default:
        return null;
    }
  };

  render() {
    return <div>{this.getFormComponent()}</div>;
  }
}

export default FormComponent;
