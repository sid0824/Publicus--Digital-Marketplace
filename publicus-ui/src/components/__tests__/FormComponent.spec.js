import toJson from 'enzyme-to-json';
import React from "react";
import {shallow, configure} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import FormComponent from "../FormComponent";
import FormConfig from '../../fixtures/FormConfig';

configure({ adapter: new Adapter() });

const mockGetFieldDecorator = jest.fn();

const mockForm = {
  getFieldDecorator: mockGetFieldDecorator
};

describe('Form Component', () => {

  beforeEach(() => {
    mockGetFieldDecorator.mockReset();
    mockGetFieldDecorator.mockReturnValue(component => component);
  });

  afterAll(() => {

  });

  test('should render text input form component', () => {
    const wrapper = shallow(
      <FormComponent
        id={1}
        question={FormConfig[1]}
        form={mockForm}
      />
    );
    expect(mockGetFieldDecorator.mock.calls[0]).toMatchSnapshot();
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('should render number input form component', () => {
    const wrapper = shallow(
      <FormComponent
        id={1}
        question={FormConfig[2]}
        form={mockForm}
      />
    );
    expect(mockGetFieldDecorator.mock.calls[0]).toMatchSnapshot();
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('should render select form component', () => {
    const wrapper = shallow(
      <FormComponent
        id={1}
        question={FormConfig[3]}
        form={mockForm}
      />
    );
    expect(mockGetFieldDecorator.mock.calls[0]).toMatchSnapshot();
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('should render switch form component', () => {
    const wrapper = shallow(
      <FormComponent
        id={1}
        question={FormConfig[4]}
        form={mockForm}
      />
    );
    expect(mockGetFieldDecorator.mock.calls[0]).toMatchSnapshot();
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('should render time picker form component', () => {
    const wrapper = shallow(
      <FormComponent
        id={1}
        question={FormConfig[5]}
        form={mockForm}
      />
    );
    expect(mockGetFieldDecorator.mock.calls[0]).toMatchSnapshot();
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('should render date picker form component', () => {
    const wrapper = shallow(
      <FormComponent
        id={1}
        question={FormConfig[6]}
        form={mockForm}
      />
    );
    expect(mockGetFieldDecorator.mock.calls[0]).toMatchSnapshot();
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('should render radio form component', () => {
    const wrapper = shallow(
      <FormComponent
        id={1}
        question={FormConfig[7]}
        form={mockForm}
      />
    );
    expect(mockGetFieldDecorator.mock.calls[0]).toMatchSnapshot();
    expect(toJson(wrapper)).toMatchSnapshot();
  });

});