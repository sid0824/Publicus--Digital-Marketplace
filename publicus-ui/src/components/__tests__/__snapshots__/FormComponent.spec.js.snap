// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form Component should render date picker form component 1`] = `
Array [
  1,
  Object {
    "rules": Array [],
  },
]
`;

exports[`Form Component should render date picker form component 2`] = `
<div>
  <FormItem
    hasFeedback={false}
  >
    <PickerWrapper
      format="YYYY-MM-DD HH:mm:ss"
      locale={Object {}}
      onChange={[Function]}
      onOk={[Function]}
      onOpenChange={[Function]}
      popupStyle={Object {}}
      showTime={true}
      transitionName="slide-up"
    />
  </FormItem>
</div>
`;

exports[`Form Component should render number input form component 1`] = `
Array [
  1,
  Object {
    "initialValue": 0,
    "rules": Array [
      Object {
        "message": "This field is compulsory",
        "required": true,
      },
    ],
  },
]
`;

exports[`Form Component should render number input form component 2`] = `
<div>
  <FormItem
    hasFeedback={false}
  >
    <InputNumber
      max={100}
      min={0}
      step={1}
      style={
        Object {
          "width": 500,
        }
      }
    />
  </FormItem>
</div>
`;

exports[`Form Component should render radio form component 1`] = `
Array [
  1,
  Object {
    "rules": Array [],
  },
]
`;

exports[`Form Component should render radio form component 2`] = `
<div>
  <FormItem
    hasFeedback={false}
  >
    <RadioGroup
      buttonStyle="outline"
    >
      <Radio
        type="radio"
        value="1"
      >
        1
      </Radio>
      <Radio
        type="radio"
        value="2"
      >
        2
      </Radio>
      <Radio
        type="radio"
        value="3"
      >
        3
      </Radio>
    </RadioGroup>
  </FormItem>
</div>
`;

exports[`Form Component should render select form component 1`] = `
Array [
  1,
  Object {
    "rules": Array [
      Object {
        "message": "This field is compulsory",
        "required": true,
      },
    ],
  },
]
`;

exports[`Form Component should render select form component 2`] = `
<div>
  <FormItem
    hasFeedback={false}
  >
    <Select
      choiceTransitionName="zoom"
      showSearch={false}
      style={
        Object {
          "width": 500,
        }
      }
      transitionName="slide-up"
    >
      <Option
        key="1"
      >
        1
      </Option>
      <Option
        key="2"
      >
        2
      </Option>
    </Select>
  </FormItem>
</div>
`;

exports[`Form Component should render switch form component 1`] = `
Array [
  1,
  Object {
    "rules": Array [],
  },
]
`;

exports[`Form Component should render switch form component 2`] = `
<div>
  <FormItem
    hasFeedback={false}
  >
    <Switch />
  </FormItem>
</div>
`;

exports[`Form Component should render text input form component 1`] = `
Array [
  1,
  Object {
    "rules": Array [
      Object {
        "message": "Value must be a email",
        "type": "email",
      },
      Object {
        "message": "This field is compulsory",
        "required": true,
      },
    ],
  },
]
`;

exports[`Form Component should render text input form component 2`] = `
<div>
  <FormItem
    hasFeedback={false}
  >
    <Input
      style={
        Object {
          "width": 500,
        }
      }
      type="text"
    />
  </FormItem>
</div>
`;

exports[`Form Component should render time picker form component 1`] = `
Array [
  1,
  Object {
    "rules": Array [
      Object {
        "message": "This field is compulsory",
        "required": true,
      },
    ],
  },
]
`;

exports[`Form Component should render time picker form component 2`] = `
<div>
  <FormItem
    hasFeedback={false}
  >
    <TimePicker
      align={
        Object {
          "offset": Array [
            0,
            -2,
          ],
        }
      }
      focusOnOpen={true}
      hideDisabledOptions={false}
      placement="bottomLeft"
      transitionName="slide-up"
    />
  </FormItem>
</div>
`;
